### PR TITLE
Add optional authorization to `dwave setup`

### LINKED ISSUE #591: Add Leap auth flow to `dwave setup`

### PR DESCRIPTION
**Pull request by randomir**
We add `--auth`, `--project` and `--oob` options to `dwave setup`.

```
$ dwave setup --help
Usage: dwave setup [OPTIONS]

  Setup optional Ocean packages and configuration file(s).

  Equivalent to running `dwave install [--all]`, followed by an optional
  `dwave auth login [--oob]` and then by `dwave config create [--full]
  [--auto-token]`.

Options:
  -a, --install-all, --all  Install all non-open-source packages available and
                            accept licenses without prompting
  --auth                    Authorize Ocean to access Leap API on user's
                            behalf. Implies --auto-token during 'dwave config
                            create' and it's mutually exclusive with --full.
  --project TEXT            Leap project for which SAPI token is pulled.
                            Defaults to active project.
  --oob                     Same as '--auth', but using OAuth out-of-band
                            flow. Use when 'localhost' not available in your
                            browser.
  --full                    Configure non-essential options manually (such as
                            endpoint and solver).
  -v, --verbose             Increase output verbosity (additive, up to 4
                            times)  [default: 0]
  --help                    Show this message and exit.
```

We now also skip contrib package(s) install/prompt if all packages are already installed.

```
$ dwave setup --auth
Optionally install non-open-source packages and configure your environment.

All optional packages already installed.

Authorizing Leap access.

Valid token found, skipping authorization.

Creating the D-Wave configuration file.

Using the simplified configuration flow.
Try 'dwave config create --full' for more options.

Updating existing configuration file: /home/user/.config/dwave/dwave.conf
Available profiles: defaults, prod, test
Updating existing profile: defaults
Fetched SAPI token for project 'Project' (PROJ) from Leap API.
Configuration saved.
```

Close #591.

### CODE REVIEW

File: dwave/cloud/cli.py [hunk 1]
```diff
@@ -904,15 +904,26 @@ def _install_contrib_package(name, verbose=0, prompt=True):
 @click.option('--install-all', '--all', '-a', default=False, is_flag=True,
               help='Install all non-open-source packages '\
                    'available and accept licenses without prompting')
+@click.option('--auth', default=False, is_flag=True,
+              help="Authorize Ocean to access Leap API on user's behalf. "
+              "Implies --auto-token during 'dwave config create' and it's "
+              "mutually exclusive with --full.")
```
Comment (JoelPasvolsky) [inline dwave/cloud/cli.py:910]: Perhaps something like "Includes running `dwave config create --auto-token`. Cannot be used together with the `--full` option."?

File: dwave/cloud/cli.py [hunk 2]
```diff
@@ -904,15 +904,26 @@ def _install_contrib_package(name, verbose=0, prompt=True):
...
+@click.option('--project', 'project', default=None,
+              help='Leap project for which SAPI token is pulled. Defaults to active project.')
```
Comment (JoelPasvolsky) [inline dwave/cloud/cli.py:912]:
```suggestion
              help='Leap project for which SAPI token is retrieved. Defaults to active project.')
```

File: dwave/cloud/cli.py [hunk 3]
```diff
@@ -904,15 +904,26 @@ def _install_contrib_package(name, verbose=0, prompt=True):
...
+@click.option('--oob', default=False, is_flag=True,
+              help="Same as '--auth', but using OAuth out-of-band flow. "
+              "Use when 'localhost' not available in your browser.")
```
Comment (JoelPasvolsky) [inline dwave/cloud/cli.py:915]:
```suggestion
              "Use when 'localhost' is not available in your browser.")
```

File: dwave/cloud/cli.py [hunk 4]
```diff
@@ -904,15 +904,26 @@ def _install_contrib_package(name, verbose=0, prompt=True):
...
 @click.option('--full', 'ask_full', default=False, is_flag=True,
-              help='Configure non-essential options (such as endpoint and solver).')
+              help='Configure non-essential options manually (such as endpoint and solver).')
```
Comment (JoelPasvolsky) [inline dwave/cloud/cli.py:917]:
```suggestion
              help='Manually configure non-essential options such as endpoint and solver.')
```

File: dwave/cloud/cli.py [hunk 5]
```diff
@@ -950,20 +976,32 @@ def auth():
...
+            return
+
     if oob:
         flow.run_oob_flow(open_browser=True)
     else:
         flow.run_redirect_flow(open_browser=True)
 
-    click.echo('Authorization completed successfully. '
-               'You can now use "dwave auth get" to fetch your token.')
+    output('Authorization completed successfully. '
```
Comment (arcondello) [inline dwave/cloud/cli.py:1003]: I am a bit confused about where we use `output()` vs `click.echo()`. For instance in `setup()` they are all `click.echo()`.

**Review by arcondello — APPROVED** [review]
One thing I am confused about, but otherwise LGTM

File: dwave/cloud/cli.py [hunk 6]
```diff
@@ -904,15 +904,26 @@ def _install_contrib_package(name, verbose=0, prompt=True):
...
               help='Increase output verbosity (additive, up to 4 times)')
-def setup(install_all, ask_full, verbose):
+@standardized_output
+def setup(install_all, auth, project, oob, ask_full, verbose, output):
     """Setup optional Ocean packages and configuration file(s).
 
     Equivalent to running `dwave install [--all]`, followed by
-    `dwave config create [--full]`.
+    an optional `dwave auth login --skip-valid [--oob]` and then by
+    `dwave config create [--full] [--auto-token]`.
```
Comment (JoelPasvolsky) [inline dwave/cloud/cli.py:926]: Why the square brackets here? You are either setting those options or not

File: releasenotes/notes/add-auth-to-dwave-setup-a0a6804a0f730851.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+---
+features:
+  - |
+    Add optional Leap authorization step to ``dwave setup`` CLI. During setup,
+    we now install optional contributed packages, authorize Ocean to access Leap
+    (``--auth`` and ``--oob`` flags), and finally create the configuration file,
+    optionally pulling SAPI token from Leap API (if authorized) for the currently
```
Comment (JoelPasvolsky) [inline releasenotes/notes/add-auth-to-dwave-setup-a0a6804a0f730851.yaml:7]:
```suggestion
    optionally retrieving SAPI token from Leap API (if authorized) for the currently
```

File: releasenotes/notes/add-skip-valid-auth-cli-option-901043cdc6f47718.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    Add ``--skip-valid`` option to ``dwave auth login`` CLI. When used,
+    authorization flow will be short-circuited if Leap access token is valid, or
```
Comment (JoelPasvolsky) [inline releasenotes/notes/add-skip-valid-auth-cli-option-901043cdc6f47718.yaml:5]:
```suggestion
    authorization flow will be skipped if Leap access token is valid or
```

**Review by JoelPasvolsky — APPROVED** [review]

File: dwave/cloud/cli.py (hunk 5 above)
Comment (randomir) [reply dwave/cloud/cli.py:1003]: Something we need to clean-up. `output` is provided by `@standardized_output`, and it handles plain text output and JSON output. But `setup` predates all that, so needs updating. I probably missed it in other places as well, like in login.

Generally, it doesn't make much sense to use it in interactive commands, like auth login is.

**Pull request merged**